eclass/databases.eclass

```diff
@@ -0,0 +1,429 @@
+#
+# - dev-db/mongodb (via "emongod" helper)
+#
+# - dev-db/postgresql[server] (via "epostgresql" helper)
+#
+# - dev-db/redis (via "eredis" helper)
+#
+# - net-misc/memcached (via "ememcached" helper)
+#
+# - virtual/mysql[server] (via "emysql" helper)
+#
+# @SUBSECTION Helper usage
+#
+# --start
+#
+# 	start the server on the default port
+#
+# --start <port>
+#
+# 	start the server on the given port
+#
+# --stop
+#
+# 	stop the server
+#
+# get-depend
+#
+# 	print a dependency string (to be included in BDEPEND)
+#
+# get-dbpath
+#
+# 	print the directory where the server stores database files
+#
+# get-logfile
+#
+#	print the path to the server's log file
+#
+# get-pidfile
+#
+#	print the path to the server's PID file
+#
+# get-sockdir
+#
+# 	print the directory where the server's sockets are located
+#
+# get-sockfile
+#
+#	print the path to the server's socket file
+
+case ${EAPI} in
+	8) ;;
+	*) die "${ECLASS}: EAPI ${EAPI} unsupported."
+esac
+
+if [[ ! ${_DATABASES_ECLASS} ]]; then
+_DATABASES_ECLASS=1
+
+# ==============================================================================
+# GENERIC FUNCTIONS
+# ==============================================================================
+
+# @FUNCTION: _databases_gen_depend
+# @USAGE: <funcname>
+# @DESCRIPTION:
+# Get a dependency string for the given helper function.
+_databases_gen_depend() {
+	local srvname=${1:1}
+	case ${srvname} in
+		memcached)
+			echo "net-misc/memcached"
+			;;
+		mongod)
+			echo "dev-db/mongodb"
+			;;
+		mysql)
+			echo "virtual/mysql[server]"
+			;;
+		postgresql)
+			echo "dev-db/postgresql[server]"
+			;;
+		redis)
+			echo "dev-db/redis"
+			;;
+		*)
+			die "${ECLASS}: unknown database: ${srvname}"
+	esac
+}
+
+# @FUNCTION: _databases_stop_service
+# @USAGE: <funcname>
+# @DESCRIPTION:
+# Default function to stop servers.  Reads PID from a file and sends the TERM
+# signal.
+_databases_stop_service() {
+	local funcname=${1?}
+	local srvname=${funcname:1}
+	local pidfile="$(${funcname} get-pidfile)"
+
+	ebegin "Stopping ${srvname}"
+	kill "$(<"${pidfile}")"
+	eend $? || die -n
+}
+
+# @FUNCTION: _databases_dispatch
+# @USAGE: <funcname> <cmd> [port]
+# @INTERNAL
+# @DESCRIPTION:
+# Start or stop a service.
+_databases_dispatch() {
+	local funcname=${1?}
+	local cmd=${2?}
+	local port=${3}
+
+	local start_fn=( _${funcname}_start ${port} )
+	if ! declare -f "${start_fn[0]}" >/dev/null; then
+		die "${ECLASS}: function not declared: ${start_fn[0]}"
+	fi
+
+	local stop_fn=( _databases_stop_service ${funcname} )
+	if declare -f _${funcname}_stop >/dev/null; then
+		stop_fn=( _${funcname}_stop )
+	fi
+
+	case ${cmd} in
+		get-depend)
+			_databases_gen_depend ${funcname}
+			;;
+		get-dbpath)
+			echo "${T}"/db/${funcname}/
+			;;
+		get-logfile)
+			echo "${T}"/db_log/${funcname}
+			;;
+		get-pidfile)
+			echo "${T}"/run/${funcname}.pid
+			;;
+		get-sockdir)
+			echo "${T}"/run/${funcname}/
+			;;
+		get-sockfile)
+			echo "${T}"/run/${funcname}.sock
+			;;
```

Comment:
How about putting them all in a single directory, e.g. `"${T}/${funcname}"/...`?